### PR TITLE
Metropolis aggregate PR

### PR DESCRIPTION
**Pull request by karalabe**
Todos:

 * ~~**[EIP 96](https://github.com/ethereum/EIPs/issues/96)**:  Blockhash refactoring~~

Done:

 * [x] **[EIP 98](https://github.com/ethereum/EIPs/issues/98)**: Removal of intermediate state roots from receipts.
 * [x] **[EIP 100](https://github.com/ethereum/EIPs/issues/100)**: Change difficulty adjustment to target mean block time including uncles.
 * [x] **[EIP 197](https://github.com/ethereum/EIPs/issues/197)**: Precompiled contracts for pairing function check.
 * [x] **[EIP 198](https://github.com/ethereum/EIPs/issues/198)**: Precompiled contract for bigint modular exponentiation.
 * [ ] **[EIP 206 (140)](https://github.com/ethereum/EIPs/issues/206)**: New opcode: REVERT ("cheap throw").
 * ~~**[EIP 208 (86)](https://github.com/ethereum/EIPs/issues/208)**: Abstraction of transaction origin and signature.~~
 * [x] **[EIP 211](https://github.com/ethereum/EIPs/issues/211)**: New opcodes: RETURNDATASIZE and RETURNDATACOPY.
 * [x] **[EIP 213 (196)](https://github.com/ethereum/EIPs/issues/213)**: Precompiled contracts for addition and scalar multiplication on the elliptic curve alt_bn128.
 * [x] **[EIP 214](https://github.com/ethereum/EIPs/issues/214)**: New opcode STATICCALL.

Separated out:

 * **[EIP 98](https://github.com/ethereum/EIPs/issues/98)**: https://github.com/ethereum/go-ethereum/pull/14750
 * **[EIP 100](https://github.com/ethereum/EIPs/issues/100)**: https://github.com/ethereum/go-ethereum/pull/14733
 * **[EIP 197](https://github.com/ethereum/EIPs/issues/197)**, **[EIP 198](https://github.com/ethereum/EIPs/issues/198)**, **[EIP 213 (196)](https://github.com/ethereum/EIPs/issues/213)**: https://github.com/ethereum/go-ethereum/pull/14959
 * **[EIP 214](https://github.com/ethereum/EIPs/issues/214)**: https://github.com/ethereum/go-ethereum/pull/14978
 * **[EIP 211](https://github.com/ethereum/EIPs/issues/211)**: https://github.com/ethereum/go-ethereum/pull/14981
 * **[EIP 206 (140)](https://github.com/ethereum/EIPs/issues/206)**: https://github.com/ethereum/go-ethereum/pull/14983

### CODE REVIEW

**Comment by holiman** [comment]
Not a concern pre-metro, but the receipt address here seems to use old nonce-based algo: https://github.com/ethereum/go-ethereum/pull/14726/files#diff-5b5942d1b7feac47d38b879796555098R121

File: core/types/receipt.go [hunk 1]
```diff
@@ -56,23 +69,28 @@ func NewReceipt(root []byte, cumulativeGasUsed *big.Int) *Receipt {
...
-	if err := s.Decode(&receipt); err != nil {
+	raw, err := s.Raw()
+	if err != nil {
 		return err
 	}
-	r.PostState, r.CumulativeGasUsed, r.Bloom, r.Logs = receipt.PostState, receipt.CumulativeGasUsed, receipt.Bloom, receipt.Logs
-	return nil
+	var hr homesteadReceiptRLP
+	var mr metropolisReceiptRLP
+	if err = rlp.DecodeBytes(raw, &mr); err == nil {
```
Comment (holiman) [inline core/types/receipt.go]: This will be active immediately. Is there a possibility that an attacker can insert a metro-receipt before metro is active?

**Comment by karalabe** [comment]
Yeah, I'm waiting for today's meeting about the address scheme before I go through the code and update everything.

**Comment by holiman** [comment]
Also not a concern pre-metro: in order for a tx to be a valid eip-86 tx, these must also be `0`: `nonce`, `value`, `gasPrice`. Is that checked anywhere? I would have expected the new signer to check it.

File: core/vm/contracts.go [hunk 2]
```diff
@@ -135,3 +150,189 @@ func (c *dataCopy) RequiredGas(input []byte) uint64 {
...
+// pairing implements a pairing pre-compile for the bn256 curve
+type pairing struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *pairing) RequiredGas(input []byte) uint64 {
+	//return 0 // TODO
+	return uint64(60000*len(input) + 40000)
```
Comment (holiman) [inline core/vm/contracts.go]: This is wrong (and it's my fault). Should be like this: https://github.com/obscuren/go-ethereum/pull/9/commits/90010b24b3dfd43574a71d353eb1a0a980696574  . @pirapira please verify?

File: core/vm/instructions.go [hunk 3]
```diff
@@ -570,6 +587,71 @@ func opCreate(pc *uint64, evm *EVM, contract *Contract, memory *Memory, stack *S
...
+		return common.BytesToAddress(crypto.Keccak256(contract.Address().Bytes(), salt.Bytes(), crypto.Keccak256(input))[12:])
+	}
+
+	contract.UseGas(gas)
+	_, addr, returnGas, suberr := evm.Create(contract, createAddress, input, gas, value)
+	// Push item on the stack based on the returned error. If the ruleset is
+	// homestead we must check for CodeStoreOutOfGasError (homestead only
+	// rule) and treat as an error, if the ruleset is frontier we must
+	// ignore this error and pretend the operation was successful.
+	if evm.ChainConfig().IsHomestead(evm.BlockNumber) && suberr == ErrCodeStoreOutOfGas {
```
Comment (holiman) [inline core/vm/instructions.go]: (Not a concern pre-metro)
Does it make sense to have special handling for Homestead, for an op only on Metro ?

File: core/vm/jump_table.go [hunk 4]
```diff
@@ -810,56 +848,73 @@ func NewFrontierInstructionSet() [256]operation {
...
-			writes:        true,
+			execute:          opCreate,
+			gasCost:          gasCreate,
+			validateStack:    makeStackFunc(3, 1),
+			memorySize:       memoryCreate,
+			valid:            true,
+			writes:           true,
+			clearsReturndata: true,
+		},
+		CREATE2: {
```
Comment (holiman) [inline core/vm/jump_table.go]: Whoa - why is this in homestead ??

**Comment by holiman** [comment]
Benchmarks : 
```
BenchmarkPrecompiledEcdsa/-Gas=3000-8  	   50000	    174450 ns/op
BenchmarkPrecompiledSha256/128-Gas=108-8         	10000000	       690 ns/op
BenchmarkPrecompiledRipeMD/128-Gas=1080-8        	 3000000	      2356 ns/op
BenchmarkPrecompiledIdentity/128-Gas=27-8        	500000000	        18.7 ns/op
BenchmarkPrecompiledModexp/eip_example1-Gas=4197376-8         	   50000	    139575 ns/op
BenchmarkPrecompiledModexp/eip_example2-Gas=4459488-8         	 1000000	      7918 ns/op
BenchmarkPrecompiledBn256Add/chfast1-Gas=0-8                  	  100000	     72309 ns/op
BenchmarkPrecompiledBn256Add/chfast2-Gas=0-8                  	  100000	     77082 ns/op
BenchmarkPrecompiledBn256ScalarMul/chfast1-Gas=0-8            	   10000	    707198 ns/op
BenchmarkPrecompiledBn256ScalarMul/chfast2-Gas=0-8            	    3000	   2748802 ns/op
BenchmarkPrecompiledBn256ScalarMul/chfast3-Gas=0-8            	    3000	   2722783 ns/op
BenchmarkPrecompiledPairing/jeff1-Gas=160000-8                	     300	  26254395 ns/op
BenchmarkPrecompiledPairing/jeff2-Gas=160000-8                	     300	  26270137 ns/op
BenchmarkPrecompiledPairing/jeff3-Gas=160000-8                	     300	  26865023 ns/op
BenchmarkPrecompiledPairing/jeff4-Gas=220000-8                	     200	  33845527 ns/op
BenchmarkPrecompiledPairing/jeff5-Gas=220000-8                	     200	  33726438 ns/op
BenchmarkPrecompiledPairing/jeff6-Gas=160000-8                	     300	  26327948 ns/op
BenchmarkPrecompiledPairing/empty_data-Gas=40000-8            	500000000	        15.6 ns/op
BenchmarkPrecompiledPairing/one_point-Gas=100000-8            	     500	  18404827 ns/op
BenchmarkPrecompiledPairing/two_point_match_2-Gas=160000-8    	     500	  16035837 ns/op
BenchmarkPrecompiledPairing/two_point_match_3-Gas=160000-8    	     300	  26066772 ns/op
BenchmarkPrecompiledPairing/two_point_match_4-Gas=160000-8    	     300	  25983786 ns/op
BenchmarkPrecompiledPairing/ten_point_match_1-Gas=640000-8    	     100	  70304992 ns/op
BenchmarkPrecompiledPairing/ten_point_match_2-Gas=640000-8    	     100	  82823341 ns/op
BenchmarkPrecompiledPairing/ten_point_match_3-Gas=160000-8    	     300	  25751592 ns/op
BenchmarkOpAdd64-8                                            	30000000	       240 ns/op
BenchmarkOpAdd128-8                                           	30000000	       248 ns/op
BenchmarkOpAdd256-8                                           	30000000	       262 ns/op
BenchmarkOpSub64-8                                            	30000000	       261 ns/op
BenchmarkOpSub128-8                                           	20000000	       305 ns/op
BenchmarkOpSub256-8                                           	20000000	       387 ns/op
BenchmarkOpMul-8                                              	20000000	       478 ns/op
BenchmarkOpDiv256-8                                           	10000000	       716 ns/op
BenchmarkOpDiv128-8                                           	20000000	       332 ns/op
BenchmarkOpDiv64-8                                            	30000000	       275 ns/op
BenchmarkOpSdiv-8                                             	10000000	      1056 ns/op
BenchmarkOpMod-8                                              	10000000	       688 ns/op
BenchmarkOpSmod-8                                             	10000000	      1138 ns/op
BenchmarkOpExp-8                                              	  200000	     33435 ns/op
BenchmarkOpSignExtend-8                                       	20000000	       428 ns/op
BenchmarkOpLt-8                                               	20000000	       396 ns/op
BenchmarkOpGt-8                                               	20000000	       392 ns/op
BenchmarkOpSlt-8                                              	10000000	       650 ns/op
BenchmarkOpSgt-8                                              	10000000	       644 ns/op
BenchmarkOpEq-8                                               	20000000	       376 ns/op
BenchmarkOpAnd-8                                              	20000000	       355 ns/op
BenchmarkOpOr-8                                               	20000000	       359 ns/op
BenchmarkOpXor-8                                              	20000000	       368 ns/op
BenchmarkOpByte-8                                             	20000000	       354 ns/op
BenchmarkOpAddmod-8                                           	10000000	       919 ns/op
BenchmarkOpMulmod-8                                           	10000000	      1186 ns/op
```

**Comment by mattdf** [comment]
Important thing to note for the benchmarks, the bn256 Go implementation from the Google crypto libs is incomplete. It's doesn't handle the case where the points you're trying to add are equal to each other.

See: 

https://github.com/golang/crypto/blob/master/bn256/bn256.go#L76

The Go equivalent of the following needs to be implemented:

https://github.com/scipr-lab/libsnark/blob/master/src/algebra/curves/alt_bn128/alt_bn128_g1.cpp#L325

File: core/vm/contracts.go [hunk 5]
```diff
@@ -135,3 +150,190 @@ func (c *dataCopy) RequiredGas(input []byte) uint64 {
...
+func (c *bigModexp) RequiredGas(input []byte) uint64 {
+	// TODO reword required gas to have error reporting and convert arithmetic
+	// to uint64.
+	if len(input) < 3*32 {
+		input = append(input, make([]byte, 3*32-len(input))...)
+	}
+	var (
+		baseLen = new(big.Int).SetBytes(input[:31])
+		expLen  = math.BigMax(new(big.Int).SetBytes(input[32:64]), big.NewInt(1))
+		modLen  = new(big.Int).SetBytes(input[65:97])
```
Comment (holiman) [inline core/vm/contracts.go]: These three are wrong, should be the same as the calculations in `Run` below: 
```golang
	baseLen  = new(big.Int).SetBytes(input[:32]).Uint64()
	expLen   = new(big.Int).SetBytes(input[32:64]).Uint64()
	modLen   = new(big.Int).SetBytes(input[64:96]).Uint64()
```

Comment (holiman) [inline core/vm/contracts.go]: Furthermore, looks really odd to use `BigMax` of `expLen` and `1`.

**Comment by karalabe** [comment]
All code pulled out from this PR, and superseded by the individual PRs linked in the description. Closing this one.

**Pull request closed without merging**